### DISCUSSION TITLE
How to handle ALL Exceptions

### DISCUSSION THREAD

**Discussion by npotier** [symfony/symfony#49549] [Q&A] [answered]
Hello, 

I'm working on a Symfony Bundle in order to add an AI recommandations when an Exception occurs, in the debug page (dev. mode).

To do that, I've declared a custom ErrorController, as the documentation indicates.

_This is a working example of the result :_

![error_ok](https://user-images.githubusercontent.com/1151094/221721827-bab2d640-18cf-4bea-82a8-5728348f661d.png)

However, this controller is not called when other exception occurs, such as PHP ParseError

![error_ko](https://user-images.githubusercontent.com/1151094/221721970-75dfb2f6-919a-4368-bc7c-6f613563ae99.png)

I've tried to create multiple listeners on different events (kernel.exception for example), but none of them is called when this type of Exception (PHP syntax) occurs.

Do you know how can I collect and handle ALL exceptions that occurs to a Symfony project ?

Thank you :)

**Comment by MatTheCat** [comment]
Hello,

I guess it boils down to [`framework.handle_all_throwables`](https://symfony.com/doc/current/reference/configuration/framework.html#handle-all-throwables); can you try setting it to `true`?

**Comment by npotier** [comment, marked as answer]
Hello @MatTheCat and thank you for your help 😄 

Unfortunatelly, this parameter is not usefull. After digging in the code, i have found that it is used by the `HttpKernel` service, but it is not yet triggered when a `Parse error` occurs, as you can see in this stack trace : 

```bash
!!  ParseError {#293
!!    #message: "syntax error, unexpected token "}", expecting ";""
!!    #code: 0
!!    #file: "./src/Controller/TestController.php"
!!    #line: 17
!!    trace: {
!!      ./src/Controller/TestController.php:17 {
!!        Symfony\Component\ErrorHandler\DebugClassLoader->loadClass(string $class): void
!!        ›         return new Response(1 + 2)
!!        ›     }
!!        › }
!!      }
!!      Symfony\Component\ErrorHandler\DebugClassLoader->loadClass() {}
!!      ./vendor/symfony/config/Resource/ClassExistenceResource.php:76 { …}
!!      ./vendor/symfony/dependency-injection/ContainerBuilder.php:347 { …}
!!      ./vendor/symfony/dependency-injection/Loader/FileLoader.php:222 { …}
!!      ./vendor/symfony/dependency-injection/Loader/FileLoader.php:113 { …}
!!      ./vendor/symfony/dependency-injection/Loader/YamlFileLoader.php:677 { …}
!!      ./vendor/symfony/dependency-injection/Loader/YamlFileLoader.php:250 { …}
!!      ./vendor/symfony/dependency-injection/Loader/YamlFileLoader.php:173 { …}
!!      ./vendor/symfony/dependency-injection/Loader/YamlFileLoader.php:129 { …}
!!      ./vendor/symfony/config/Loader/FileLoader.php:163 { …}
!!      ./vendor/symfony/config/Loader/FileLoader.php:94 { …}
!!      ./vendor/symfony/dependency-injection/Loader/FileLoader.php:64 { …}
!!      ./vendor/symfony/dependency-injection/Loader/Configurator/ContainerConfigurator.php:64 { …}
!!      ./vendor/symfony/framework-bundle/Kernel/MicroKernelTrait.php:57 { …}
!!      ./vendor/symfony/framework-bundle/Kernel/MicroKernelTrait.php:179 { …}
!!      ./vendor/symfony/dependency-injection/Loader/ClosureLoader.php:36 { …}
!!      ./vendor/symfony/config/Loader/DelegatingLoader.php:37 { …}
!!      ./vendor/symfony/framework-bundle/Kernel/MicroKernelTrait.php:187 { …}
!!      ./vendor/symfony/http-kernel/Kernel.php:585 { …}
!!      ./vendor/symfony/http-kernel/Kernel.php:486 { …}
!!      ./vendor/symfony/http-kernel/Kernel.php:709 { …}
!!      ./vendor/symfony/http-kernel/Kernel.php:122 { …}
!!      ./vendor/symfony/framework-bundle/Console/Application.php:148 { …}
!!      ./vendor/symfony/framework-bundle/Console/Application.php:69 { …}
!!      ./vendor/symfony/console/Application.php:168 { …}
!!      ./vendor/symfony/runtime/Runner/Symfony/ConsoleApplicationRunner.php:54 { …}
!!      ./vendor/autoload_runtime.php:30 { …}
!!      ./bin/console:11 { …}
!!    }
!!  }
```

😅 It seems to be very complicated to add a new feature to the debug exception view to handle all the cases, but once again, thank you for your help

**Reply by MatTheCat** [reply]
Oh yeah if your controller is registered as a service (it is by default) this kind of error will occur at compile time so `HttpKernel::handle` won’t even be called and `framework.handle_all_throwables` will serve no purpose.

It looks like you will have to look into the `ErrorHandler` component. Implementing an `ErrorEnhancerInterface` seems like a good start.

**Reply by npotier** [reply]
Yes, That's what i have figured out too. The problem is that in this situation, I don't have access to services, such as the cache that I want to use, to code the logic of the AI Anwser.

Anyway, I will work on it and publish it as an Open Source Bundle when its ready. 

Thanks.

**Reply by MatTheCat** [reply]
Could you link it when it's published? I would be interested to know how you did it.

**Reply by npotier** [reply]
Hello, following up : 

https://github.com/acseo/AIErrorExplainedBundle

This Bundle proposes to override the default ErrorHandler (in dev mode) to add AI recommandation to solve the encountered error.

it is in beta, feel free to install it, test it, and give me feedback about it 🙂

**Reply by MatTheCat** [reply]
Nice did not know about `extra.runtime.error_handler`, thanks for sharing!

That being said it is weird having to duplicate Symfony’s `ErrorHandler` to achieve what you want as it seems to provide some extension points :thinking:
